src/lib/Persistence/Cache/Tag/CacheIdentifierGenerator.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\Core\Persistence\Cache\Tag;
+
+use eZ\Publish\Core\Base\Exceptions\InvalidArgumentException;
+
+/**
+ * @internal
+ */
+final class CacheIdentifierGenerator implements CacheIdentifierGeneratorInterface
+{
+    private const PLACEHOLDER = '-%s';
+
+    /** @var string */
+    private $prefix;
+
+    /** @var array<string,string> */
+    private $tagPatterns;
+
+    /** @var array<string,string> */
+    private $keyPatterns;
+
+    public function __construct(string $prefix, array $tagPatterns, array $keyPatterns)
+    {
+        $this->prefix = $prefix;
+        $this->tagPatterns = $tagPatterns;
+        $this->keyPatterns = $keyPatterns;
+    }
+
+    /**
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     */
+    public function generateTag(string $patternName, array $values = [], bool $withPrefix = false): string
+    {
+        if (!isset($this->tagPatterns[$patternName])) {
+            throw new InvalidArgumentException($patternName, 'Undefined tag pattern');
```

Comment:
Ideally the exception should help find the right pattern - possibly by providing a list of values already present, so that their source can be found.

```suggestion
            throw new InvalidArgumentException($patternName, sprintf(
                'Undefined tag pattern "%s". Known pattern names are: "%s"',
                $patternName,
                implode('", "', array_keys($this->tagPatterns))
            );
```

The same applies to the code for generating keys.